### PR TITLE
[proof-new] Adds a proof-producing CNF converter

### PR DESCRIPTION
**Pull request by HanielB**
A proof generator that wraps the original CNF stream, to be used when the prop engine is proof producing. It tracks in a lazy cdproof both the concrete clausification steps and the proof generators of the formulas being clausified (in particular, theory lemmas).

### CODE REVIEW

File: src/prop/proof_cnf_stream.cpp [hunk 1]
```diff
@@ -0,0 +1,1265 @@
...
+    default:
+    {
+      // negate
+      Node nnode = negated ? node.negate() : static_cast<Node>(node);
+      // Atoms
+      SatLiteral lit = toCNF(node, negated);
+      bool added = d_cnfStream.assertClause(nnode, lit);
+      if (negated && added && nnode != node.notNode())
+      {
+        d_proof.addStep(
```
Comment (ajreynol) [inline src/prop/proof_cnf_stream.cpp:126]: Can you explain whats going on here?

File: src/prop/proof_cnf_stream.cpp [hunk 2]
```diff
@@ -0,0 +1,1265 @@
...
+      {
+        disjuncts.push_back(node[i].notNode());
+      }
+      Node clauseNode = NodeManager::currentNM()->mkNode(kind::OR, disjuncts);
+      d_proof.addStep(clauseNode, PfRule::NOT_AND, {node.notNode()}, {});
+      Trace("cnf") << "ProofCnfStream::convertAndAssertAnd: NOT_AND added "
+                   << clauseNode << "\n";
+      // justify normalized clause as well, since that's what will be saved in
+      // the SAT solver and registered with prop engine
+      Node normClauseNode = d_psb.factorReorderElimDoubleNeg(clauseNode);
```
Comment (ajreynol) [inline src/prop/proof_cnf_stream.cpp:185]: It seems that there is an invariant that this method should be called whenever you registerSatAssumptions below.

Would it make sense to make a utility function that does this?  It seems this pattern is repeated several times below.

File: src/prop/proof_cnf_stream.cpp [hunk 3]
```diff
@@ -0,0 +1,1265 @@
...
+                   << clauseNode << "\n";
+      Node normClauseNode = d_psb.factorReorderElimDoubleNeg(clauseNode);
+      if (Trace.isOn("cnf") && normClauseNode != clauseNode)
+      {
+        Trace("cnf")
+            << "ProofCnfStream::convertAndAssertIff: steps to normalized "
+            << normClauseNode << "\n";
+      }
+      MinisatSatSolver* minisat =
+          static_cast<MinisatSatSolver*>(d_cnfStream.d_satSolver);
```
Comment (ajreynol) [inline src/prop/proof_cnf_stream.cpp:381]: Should you just do a single static_cast in the constructor?  Or better, have the constructor of this class take `MinisatSatSolver *`?

**Review by ajreynol — CHANGES_REQUESTED** [review]
A few initial comments.

File: src/prop/proof_cnf_stream.cpp (hunk 1 above)
Comment (HanielB) [reply src/prop/proof_cnf_stream.cpp:126]: It's tracking double negation elimination. I'll add a comment.

File: src/prop/proof_cnf_stream.cpp (hunk 3 above)
Comment (HanielB) [reply src/prop/proof_cnf_stream.cpp:381]: I will actually make this class take the SatProofManager, that's the real dependency here.

File: src/prop/proof_cnf_stream.cpp (hunk 2 above)
Comment (HanielB) [reply src/prop/proof_cnf_stream.cpp]: Indeed. I've updated the class to do this.

**Review by ajreynol — APPROVED** [review]
LGTM

**Pull request merged**